Handle multiple station URLs on Register and more

DaniElectra opened the pull request:
- Games like Mario Kart 7 or Mario Tennis Open set the local station and
the public station. In these cases, use the public station given by the client.
- Add the PID and RVCID on both stations, as they are required on NAT
traversal and not all games will call `ReportNATProperties` to add them.
- Implement `MatchmakeExtension::CloseParticipation`
- and more

DaniElectra committed:
Handle multiple station URLs on Register

Games like Mario Kart 7 or Mario Tennis Open set two station URLs on the
Register method, the local sattion and the public station. In these
cases, use the public station given by the client.

Also add the PID and RVCID on both stations, as they are required on NAT
traversal and not all games will call `ReportNATProperties` to add them.

DaniElectra committed:
Use sequential gathering IDs

DaniElectra committed:
Implement CloseParticipation

Called when a gathering fills up.

DaniElectra committed:
Various fixes

jonbarrow commented on globals/matchmaking_globals.go:
```diff
@@ -12,3 +12,4 @@ type CommonMatchmakeSession struct {
 }
 
 var Sessions map[uint32]*CommonMatchmakeSession
+var CurrentGatheringID uint32
```
Not a massive deal but `nex-go` already provides a `Counter` struct which is used internally there for things like client connection IDs and packet sequence IDs, but can also just be reused here

jonbarrow commented on secure-connection/register.go:
```diff
@@ -17,15 +17,27 @@ func register(err error, client *nex.Client, callID uint32, stationUrls []*nex.S
-	publicStation.SetType(3)
+	// * A NEX client can set the public station URL by setting two URLs on the array
+	var publicStation *nex.StationURL
+	if len(stationUrls) > 1 {
+		publicStation = stationUrls[1]
+	} else {
```
Public stations seem to use a `type` parameter of 3. This feels like a safer way to check for the public station rather than assuming it's the 2nd value in the list

Out of curiosity, how does it _know_ it's public station here?

DaniElectra replied:
> Out of curiosity, how does it know it's public station here?

I'm not sure tbh. It seems to be a thing external to the server though, because it also uses the public IP even when the server is on the same network as the client

jonbarrow commented on secure-connection/register.go:
```diff
@@ -17,15 +17,27 @@ func register(err error, client *nex.Client, callID uint32, stationUrls []*nex.S
+		publicStation.SetType(3)
+	}
+
+	localStation.SetPID(client.PID())
 	publicStation.SetPID(client.PID())
 
+	localStation.SetRVCID(client.ConnectionID())
+	publicStation.SetRVCID(client.ConnectionID())
+
```
Has this been tested on a wide number of games? I tried always setting these values (PID and connection ID) on all the stations in `GetSessionURLs` while working on Minecraft, and it seemed to break PIA in Mario Kart 8. It could have been a fluke, but I tried a few times and the guest couldn't connect to the host until I reverted those changes

DaniElectra replied:
> I tried always setting these values (PID and connection ID) on all the stations in GetSessionURLs while working on Minecraft, and it seemed to break PIA in Mario Kart 8.

That's weird. AFAIK, the PID and connection ID should be set on the stations anyway when calling `ReportNATProperties`. We can't rely on that for NEX 2.x games though, because they don't seem to call that method at any time, so I'm setting them on `Register` aswell.

jonbarrow replied:
I'll merge this for now, and check Mario Kart 8 again. If it breaks, then we have to find a different solution

DaniElectra committed:
Use NEX Counter for gathering IDs

DaniElectra committed:
Search for public station on Register

This is more reliable than just taking the second URL.

DaniElectra committed:
Better check for nil clients

EndParticipation triggered a nil client during testing. Add checks for
all client searches.
